lib/Dialect/Verif/Transforms/LowerContracts.cpp

```diff
@@ -0,0 +1,109 @@
+using namespace hw;
+
+namespace {
+struct LowerContractsPass
+    : verif::impl::LowerContractsPassBase<LowerContractsPass> {
+  void runOnOperation() override;
+};
+
+template <typename FROM, typename TO>
+void replaceContractOp(PatternRewriter &rewriter, Block *body) {
+  for (auto op : llvm::make_early_inc_range(body->getOps<FROM>())) {
+    auto label = op.getLabel();
+    StringAttr labelAttr;
+    if (label) {
+      labelAttr = rewriter.getStringAttr(label.value());
+    }
+    rewriter.replaceOpWithNewOp<TO>(op, op.getProperty(), op.getEnable(),
+                                    labelAttr);
+  }
+}
+
+struct HWOpRewritePattern : public OpRewritePattern<HWModuleOp> {
+  using OpRewritePattern<HWModuleOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(HWModuleOp op,
+                                PatternRewriter &rewriter) const override {
+    auto formalOp = rewriter.create<verif::FormalOp>(
+        op.getLoc(), op.getNameAttr(), rewriter.getDictionaryAttr({}));
+
+    // Clone module body into fomal op body
+    rewriter.cloneRegionBefore(op.getRegion(), formalOp.getBody(),
+                               formalOp.getBody().end());
+
+    auto *bodyBlock = &formalOp.getBody().front();
+
+    // Erase hw.output
+    rewriter.eraseOp(bodyBlock->getTerminator());
+
+    // Convert block args to symbolic values
+    rewriter.setInsertionPointToStart(bodyBlock);
+    for (auto arg : llvm::make_early_inc_range(bodyBlock->getArguments())) {
+      auto sym =
+          rewriter.create<verif::SymbolicValueOp>(arg.getLoc(), arg.getType());
+      rewriter.replaceAllUsesWith(arg, sym);
+    }
+    bodyBlock->eraseArguments(0, bodyBlock->getNumArguments());
+
+    // Inline contract ops
+    for (auto contractOp :
+         llvm::make_early_inc_range(bodyBlock->getOps<verif::ContractOp>())) {
+
+      // Convert ensure to assert, require to assume
+      rewriter.setInsertionPointToEnd(&contractOp.getBody().front());
+      Block *contractBlock = &contractOp.getBody().front();
+      replaceContractOp<EnsureOp, AssertOp>(rewriter, contractBlock);
+      replaceContractOp<RequireOp, AssumeOp>(rewriter, contractBlock);
+
+      // Inline body
+      rewriter.inlineBlockBefore(&contractOp.getBody().front(),
+                                 &formalOp.getBody().front(),
+                                 formalOp.getBody().front().end());
+
+      // Replace results with inputs and erase
+      for (auto [input, result] :
+           llvm::zip(contractOp.getResults(), contractOp.getInputs())) {
+        rewriter.replaceAllUsesWith(input, result);
+      }
+      rewriter.eraseOp(contractOp);
+    }
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+} // namespace
+
+void LowerContractsPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
+  patterns.add<HWOpRewritePattern>(patterns.getContext());
+
+  if (failed(applyPatternsAndFoldGreedily(getOperation(), std::move(patterns))))
+    signalPassFailure();
```

Comment:
Since you are only matching on modules, you can also just have a `for` loop iterating over the modules, which would be a bit more efficient.